doc/adr/0008-jetstream-kv.md

```diff
@@ -0,0 +1,326 @@
+# 8. JetStream based Key-Value Stores
+
+|Metadata|Value|
+|--------|-----|
+|Date    |2021-06-30|
+|Author  |@ripienaar|
+|Status  |Partially Implemented|
+|Tags    |jetstream, client, kv|
+
+## Context
+
+This document describes a design and initial implementation of a JetStream backed key-value store. The initial implementation
+is available in the CLI as `nats kv` with an initial reference client package implementation in `jsm.go` repository to support
+the CLI.
+
+This document aims to guide client developers in implementing this feature in the language clients we maintain.
+
+## Overview
+
+We intend to hit a basic initial feature set as below, with some future facing goals as indicated:
+
+Initial feature list:
+
+ * Multiple named buckets full of keys with n historical values kept per value
```

Comment:
We dont support something to limit the number of keys in JS.  Versions per key is the History mentioned

Max size per value, Max size per bucket good, will add.

Reply:
Should we add max subjects to a stream definition?

Reply:
Might be interesting - then I can also get from JS a list of subjects and count of subjects - these are why I dont have easy to access things like `Keys()` because I have to walk the entire stream to answer those questions.

Reply:
OK I think this might be needed, but we can do simple math in the meantime.

Reply:
@matthiashanel yes, it will be removing a message to make space so should be allowed to write to a full bucket.